pkg/reconciler/pipelinerun/resources/pipelinerunstate.go

```diff
@@ -412,7 +412,7 @@ func (facts *PipelineRunFacts) GetFinalTasks() PipelineRunState {
 	if facts.checkDAGTasksDone() {
 		// return list of tasks with all final tasks
 		for _, t := range facts.State {
-			if facts.isFinalTask(t.PipelineTask.Name) && !t.isSuccessful() {
+			if facts.isFinalTask(t.PipelineTask.Name) && !t.isScheduled() {
```

Comment:
I think this would break retries in finally tasks - the current code purposely includes failed tasks because `getNextTasks` will do the filtering again, and it needs failed tasks to be in so that `getRetryableTasks` can get tasks which failed but may be retried. The comment on L405 should be updated to reflect that.

I find the code a bit confusing here because it seems that the filter logic is shared between the `GetFinalTasks` and `getNextTasks`. I think that L415 could be:

```go
   if facts.isFinalTask(t.PipelineTask.Name) {
```

So that we get *all* final tasks and the filtering is done only by `getNextTasks`